feat(MeasureTheory): Pullback of measure preserve properties

DavidLedvinka opened the pull request:
Add a variety of theorems that prove some property is preserved under the pullback of a measure (under appropriate sufficient conditions on the function). Some of these theorems are needed in the FLT project.

EtienneC30 commented on Mathlib/MeasureTheory/Group/Measure.lean:
```diff
@@ -753,6 +772,15 @@ theorem isHaarMeasure_map [BorelSpace G] [ContinuousMul G] {H : Type*} [Group H]
       exact IsCompact.measure_lt_top (g.isCompact_preimage_of_isClosed hK.closure isClosed_closure)
     toIsOpenPosMeasure := hf.isOpenPosMeasure_map h_surj }
 
+theorem _root_.Topology.IsOpenEmbedding.isHaarMeasure_comap [BorelSpace G] [MeasurableMul G]
```
Again I'm not sure we want these out of the `Measure` namespace.
```suggestion
protected theorem IsHaarMeasure.comap [BorelSpace G] [MeasurableMul G]
```

DavidLedvinka replied:
Is it ok that this one disagrees with the convention used for map here? (`isHaarMeasure_map`)

EtienneC30 replied:
I think it's ok. The convention for `isHaarMeasure_map` comes from the hypothesis `h_prop` which is easily obtained in the special case of a `MulEquiv`, which justifies the existence of `MulEquiv.isHaarMeasure_map` and thus the naming (although I am not sure it should have been done this way). But there isn't such a hypothesis here so you wouldn't need these extra versions so I think `IsHaarMeasure.comap` is fine.
If you are note sure about these questions about namespaces, feel free to open a discussion on Zulip to ask for other opinions, I am not pretending that what I said is a clear consensus.

EtienneC30 commented on Mathlib/MeasureTheory/Measure/OpenPos.lean:
```diff
@@ -146,6 +146,14 @@ theorem _root_.Continuous.isOpenPosMeasure_map [OpensMeasurableSpace X]
   rw [Measure.map_apply hf.measurable hUo.measurableSet]
   exact (hUo.preimage hf).measure_ne_zero μ (hf_surj.nonempty_preimage.mpr hUne)
 
+theorem _root_.Topology.IsOpenEmbedding.isOpenPosMeasure_comap [BorelSpace X]
```
```suggestion
protected theorem IsOpenPosMeasure.comap [BorelSpace X]
```

EtienneC30 commented on Mathlib/MeasureTheory/Measure/Regular.lean:
```diff
@@ -247,6 +247,20 @@ theorem map' {α β} [MeasurableSpace α] [MeasurableSpace β] {μ : Measure α}
   refine ⟨f '' K, image_subset_iff.2 hKU, hAB' _ hKc, ?_⟩
   rwa [f.map_apply, f.preimage_image]
 
+protected theorem _root_.MeasurableEmbedding.innerRegularWRT_comap
+    {α β : Type*} [MeasurableSpace α] [MeasurableSpace β]
+    {μ : Measure β} {pa qa : Set α → Prop} {pb qb : Set β → Prop}
```
Based on lemmas around this should probably just be called `comap`, without the `MeasurableEmbedding` namespace. Also `MeasurableSpace β` should be implicit.
```suggestion
protected theorem InnerRegularWRT.comap
    {α β : Type*} [MeasurableSpace α] {mβ : MeasurableSpace β}
    {μ : Measure β} {pa qa : Set α → Prop} {pb qb : Set β → Prop}
```

EtienneC30 commented on Mathlib/MeasureTheory/Measure/Regular.lean:
```diff
@@ -374,6 +388,23 @@ protected theorem map [OpensMeasurableSpace α] [MeasurableSpace β] [Topologica
   refine ⟨f.symm ⁻¹' U, image_subset_iff.1 hAU, this, ?_⟩
   rwa [map_apply f.measurable this.measurableSet, f.preimage_symm, f.preimage_image]
 
+open Topology in
```
This can be removed.
```suggestion
```

EtienneC30 commented on Mathlib/MeasureTheory/Measure/Regular.lean:
```diff
@@ -374,6 +388,23 @@ protected theorem map [OpensMeasurableSpace α] [MeasurableSpace β] [Topologica
   rwa [map_apply f.measurable this.measurableSet, f.preimage_symm, f.preimage_image]
 
+open Topology in
+protected theorem comap_of_continuous_measurableEmbedding
+    [MeasurableSpace β] [TopologicalSpace β] (μ : Measure β) [OuterRegular μ]
```
I don't think we need `protected` here. I think you need one more "of" to follow naming convention, and `MeasurableSpace β` should be implicit.
```suggestion
theorem comap_of_continuous_of_measurableEmbedding
    {mβ : MeasurableSpace β} [TopologicalSpace β] (μ : Measure β) [OuterRegular μ]
```

EtienneC30 commented on Mathlib/MeasureTheory/Measure/Regular.lean:
```diff
@@ -374,6 +388,23 @@ protected theorem map [OpensMeasurableSpace α] [MeasurableSpace β] [Topologica
+    rw [MeasurableEmbedding.comap_apply f_me]
+    apply lt_of_le_of_lt (measure_mono (Set.image_preimage_subset _ _)) hμU
+
+protected theorem comap [BorelSpace α] [MeasurableSpace β] [TopologicalSpace β] [BorelSpace β]
```
```suggestion
protected theorem comap [BorelSpace α] {mβ : MeasurableSpace β} [TopologicalSpace β] [BorelSpace β]
```

EtienneC30 commented on Mathlib/MeasureTheory/Measure/Regular.lean:
```diff
@@ -374,6 +388,23 @@ protected theorem map [OpensMeasurableSpace α] [MeasurableSpace β] [Topologica
+    {f : α → β} (f_cont : Continuous f) (f_me : MeasurableEmbedding f) :
+    (μ.comap f).OuterRegular where
+  outerRegular A hA r hr := by
+    rw [MeasurableEmbedding.comap_apply f_me] at hr
+    obtain ⟨U, hUA, Uopen, hμU⟩ :=
+      OuterRegular.outerRegular (f_me.measurableSet_image' hA) r hr
+    refine ⟨f ⁻¹' U, by rwa [Superset, ← Set.image_subset_iff], Uopen.preimage f_cont, ?_⟩
+    rw [MeasurableEmbedding.comap_apply f_me]
+    apply lt_of_le_of_lt (measure_mono (Set.image_preimage_subset _ _)) hμU
```
No need for the `Set` namespace. Also when `apply` closes the goal it's nicer to use `exact`.
```suggestion
    rw [f_me.comap_apply] at hr
    obtain ⟨U, hUA, Uopen, hμU⟩ :=
      OuterRegular.outerRegular (f_me.measurableSet_image' hA) r hr
    refine ⟨f ⁻¹' U, by rwa [Superset, ← image_subset_iff], Uopen.preimage f_cont, ?_⟩
    rw [f_me.comap_apply]
    exact lt_of_le_of_lt (measure_mono (image_preimage_subset _ _)) hμU
```

EtienneC30 commented on Mathlib/MeasureTheory/Measure/Regular.lean:
```diff
@@ -695,6 +726,22 @@ protected theorem map_iff [BorelSpace α] [MeasurableSpace β] [TopologicalSpace
   simp
 
+open Topology in
+protected theorem _root_.Topology.IsOpenEmbedding.innerRegular_comap [BorelSpace α]
```
Again I don't think we should keep them out of the `Measure` namespace. I think what is done for [MeasureTheory.Measure.InnerRegularWRT.map](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/Regular.html#MeasureTheory.Measure.InnerRegularWRT.map) works nice: you can call this one `comap'`.
```suggestion
protected theorem InnerRegular.comap' [BorelSpace α]
```

EtienneC30 commented on Mathlib/MeasureTheory/Measure/Regular.lean:
```diff
@@ -695,6 +726,22 @@ protected theorem map_iff [BorelSpace α] [MeasurableSpace β] [TopologicalSpace
 
+open Topology in
+protected theorem _root_.Topology.IsOpenEmbedding.innerRegular_comap [BorelSpace α]
+    [MeasurableSpace β] [TopologicalSpace β] [BorelSpace β]
```
```suggestion
    {mβ : MeasurableSpace β} [TopologicalSpace β] [BorelSpace β]
```

EtienneC30 commented on Mathlib/MeasureTheory/Measure/Regular.lean:
```diff
@@ -695,6 +726,22 @@ protected theorem map_iff [BorelSpace α] [MeasurableSpace β] [TopologicalSpace
+    (fun _ hU ↦ hf.measurableEmbedding.measurableSet_image' hU)
+    (fun _ hKrange hK ↦ hf.isInducing.isCompact_preimage' hK hKrange)
+
+protected theorem comap [BorelSpace α]
+    [MeasurableSpace β] [TopologicalSpace β] [BorelSpace β]
+    (μ : Measure β) [InnerRegular μ] (f : α ≃ₜ β) :
```
```suggestion
protected theorem comap [BorelSpace α] {mβ : MeasurableSpace β} [TopologicalSpace β] [BorelSpace β]
    (μ : Measure β) [InnerRegular μ] (f : α ≃ₜ β) :
```

EtienneC30 commented on Mathlib/MeasureTheory/Measure/Regular.lean:
```diff
@@ -1017,6 +1064,25 @@ protected theorem map_iff [BorelSpace α] [MeasurableSpace β] [TopologicalSpace
   simp
 
+open Topology in
+protected theorem _root_.Topology.IsOpenEmbedding.regular_comap [BorelSpace α]
+    [MeasurableSpace β] [TopologicalSpace β] [BorelSpace β] {μ : Measure β} [Regular μ]
```
```suggestion
protected theorem Regular.comap' [BorelSpace α]
    {mβ : MeasurableSpace β} [TopologicalSpace β] [BorelSpace β] {μ : Measure β} [Regular μ]
```

EtienneC30 commented on Mathlib/MeasureTheory/Measure/Regular.lean:
```diff
@@ -1017,6 +1064,25 @@ protected theorem map_iff [BorelSpace α] [MeasurableSpace β] [TopologicalSpace
+protected theorem _root_.Topology.IsOpenEmbedding.regular_comap [BorelSpace α]
+    [MeasurableSpace β] [TopologicalSpace β] [BorelSpace β] {μ : Measure β} [Regular μ]
+    (f : α → β) (hf : IsOpenEmbedding f) : (μ.comap f).Regular := by
+  haveI :=
+    OuterRegular.comap_of_continuous_measurableEmbedding
+      μ hf.continuous hf.measurableEmbedding
```
```suggestion
  haveI := OuterRegular.comap_of_continuous_measurableEmbedding
    μ hf.continuous hf.measurableEmbedding
```

EtienneC30 commented on Mathlib/MeasureTheory/Measure/Regular.lean:
```diff
@@ -1017,6 +1064,25 @@ protected theorem map_iff [BorelSpace α] [MeasurableSpace β] [TopologicalSpace
+  haveI :=
+    IsFiniteMeasureOnCompacts.comap_of_continuous_measurableEmbedding
+      μ hf.continuous hf.measurableEmbedding
+  exact
+    ⟨hf.measurableEmbedding.innerRegularWRT_comap Regular.innerRegular (μ := μ)
+      (fun _ hU ↦ hf.isOpen_iff_image_isOpen.mp hU)
+      (fun _ hKrange hK ↦ hf.isInducing.isCompact_preimage' hK hKrange)⟩
```
```suggestion
  exact ⟨hf.measurableEmbedding.innerRegularWRT_comap Regular.innerRegular (μ := μ)
    (fun _ hU ↦ hf.isOpen_iff_image_isOpen.mp hU)
    (fun _ hKrange hK ↦ hf.isInducing.isCompact_preimage' hK hKrange)⟩
```

EtienneC30 commented on Mathlib/MeasureTheory/Measure/Regular.lean:
```diff
@@ -1017,6 +1064,25 @@ protected theorem map_iff [BorelSpace α] [MeasurableSpace β] [TopologicalSpace
+      (fun _ hU ↦ hf.isOpen_iff_image_isOpen.mp hU)
+      (fun _ hKrange hK ↦ hf.isInducing.isCompact_preimage' hK hKrange)⟩
+
+protected theorem comap [BorelSpace α] [MeasurableSpace β] [TopologicalSpace β]
```
```suggestion
protected theorem comap [BorelSpace α] {mβ : MeasurableSpace β} [TopologicalSpace β]
```

EtienneC30 commented on Mathlib/MeasureTheory/Measure/Typeclasses/Finite.lean:
```diff
@@ -594,6 +594,15 @@ noncomputable irreducible_def MeasureTheory.Measure.finiteSpanningSetsInOpen' [T
   obtain ⟨n, rfl⟩ : ∃ n : ℕ, f n = t := by simpa only using tT
   exact mem_iUnion_of_mem _ xt
 
+protected theorem IsFiniteMeasureOnCompacts.comap_of_continuous_measurableEmbedding
```
I think this one should be called `comap'` and go with its homeomorph version in `Borel.Basic`. Once again `MeasurableSpace` should be implicit.

EtienneC30 replied:
Oh you're right! In this case I think you should leave it in `Typeclasses.Finite`. I think a good place is after line 351, with other preserving properties.

EtienneC30 reviewed (commented):
Thank you for all of these. My main concerns are about naming. I think all these results should remain in the `Measure` namespace, as is the case for other mapping lemmas such as [MeasureTheory.Measure.Regular.map](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/Regular.html#MeasureTheory.Measure.Regular.map). Also, all the codomains should have an implicit `MeasurableSpace` parameter instead of a typeclass argument because they come with a measure, so we need an implicit parameter to be able to use a measure which is not defined on the preferred sigma-algebra (typically when using [MeasureTheory.Measure.trim](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/Trim.html#MeasureTheory.Measure.trim)).

DavidLedvinka commented:
> Thank you for all of these. My main concerns are about naming. I think all these results should remain in the `Measure` namespace, as is the case for other mapping lemmas such as [MeasureTheory.Measure.Regular.map](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/Regular.html#MeasureTheory.Measure.Regular.map). Also, all the codomains should have an implicit `MeasurableSpace` parameter instead of a typeclass argument because they come with a measure, so we need an implicit parameter to be able to use a measure which is not defined on the preferred sigma-algebra (typically when using [MeasureTheory.Measure.trim](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/Trim.html#MeasureTheory.Measure.trim)).

I agree with your argument about the `Measure` namespace from an organizational point of view but the reason I put them in the `Topology.openEmbedding` namespace is so that you can use dot notation (and otherwise these are quite cumbersome to use). Im ok with either way but wanted to make that trade off clear. 

For the `MeasurableSpace` parameter you have mentioned this before but I apologize I didn't internalize when exactly it applies (and here just followed the surrounding theorems). Is the rule of thumb that you should make the `MeasurableSpace` parameter implicit iff the theorem takes in a measure?

EtienneC30 commented:
I suspected that you did this to allow dot notation. The thing is I think almost every lemma could be named to use dot notation on one of its parameters, but we would lose all the organization namespaces provide, so we have to draw the line somewhere. I think that having the scheme `Property.map` when `Property` is transported by a function is a good convention which makes sense in a wide variety of contexts from algebra to analysis so that's why I think we should do it like this.

Regarding the `MeasurableSpace` impliciteness, it is not completely clear to me whether this is an iff, but one direction is definitely true: if there is a measure on the space in your lemma, then the measurable space should be implicit. If you have a lemma
```lean
lemma foo {X : Type*} [MeasurableSpace X] (μ : Measure X) : ... := ...
```
In a context with
```lean
variable {X : Type*} {m m0 : MeasurableSpace X} (h : m ≤ m0) (μ : Measure X)
```
and you want to apply `foo` to `μ.trim h`, what will happen is that `foo` will synthesize `MeasurableSpace X` using the last one introduced, which is `m0`, but as `μ.trim h` is a measure over `m`, there will be a conflict which will cause an error. If you write `{mX : MeasurableSpace X}` in `foo` on the other hand, this parameter will be inferred from the type of `μ.trim h` and everything will go well.

EtienneC30 commented on Mathlib/MeasureTheory/Measure/Regular.lean:
```diff
@@ -374,6 +388,22 @@ protected theorem map [OpensMeasurableSpace α] [MeasurableSpace β] [Topologica
   refine ⟨f.symm ⁻¹' U, image_subset_iff.1 hAU, this, ?_⟩
   rwa [map_apply f.measurable this.measurableSet, f.preimage_symm, f.preimage_image]
 
+theorem comap_of_continuous_of_measurableEmbedding
```
In fact we can put `comap'` here I think.
```suggestion
protected theorem comap'
```

EtienneC30 commented on Mathlib/MeasureTheory/Measure/Regular.lean:
```diff
@@ -374,6 +388,22 @@ protected theorem map [OpensMeasurableSpace α] [MeasurableSpace β] [Topologica
+    (μ.comap f).OuterRegular where
+  outerRegular A hA r hr := by
+    rw [f_me.comap_apply] at hr
+    obtain ⟨U, hUA, Uopen, hμU⟩ :=
+      OuterRegular.outerRegular (f_me.measurableSet_image' hA) r hr
```
```suggestion
    obtain ⟨U, hUA, Uopen, hμU⟩ := OuterRegular.outerRegular (f_me.measurableSet_image' hA) r hr
```

EtienneC30 commented on Mathlib/MeasureTheory/Measure/Regular.lean:
```diff
@@ -374,6 +388,22 @@ protected theorem map [OpensMeasurableSpace α] [MeasurableSpace β] [Topologica
+      OuterRegular.outerRegular (f_me.measurableSet_image' hA) r hr
+    refine ⟨f ⁻¹' U, by rwa [Superset, ← image_subset_iff], Uopen.preimage f_cont, ?_⟩
+    rw [f_me.comap_apply]
+    exact lt_of_le_of_lt (measure_mono (image_preimage_subset _ _)) hμU
```
Just mentioning this other spelling, but feel free to keep yours.
```suggestion
    exact (measure_mono (image_preimage_subset _ _)).trans_lt hμU
```

EtienneC30 commented on Mathlib/MeasureTheory/Measure/Regular.lean:
```diff
@@ -695,6 +725,21 @@ protected theorem map_iff [BorelSpace α] [MeasurableSpace β] [TopologicalSpace
+    {μ : Measure β} [H : InnerRegular μ] {f : α → β} (hf : IsOpenEmbedding f) :
+    (μ.comap f).InnerRegular where
+  innerRegular :=
+    InnerRegularWRT.comap H.innerRegular hf.measurableEmbedding
```
```suggestion
    H.innerRegular.comap hf.measurableEmbedding
```

EtienneC30 commented on Mathlib/MeasureTheory/Measure/Regular.lean:
```diff
@@ -1017,6 +1062,22 @@ protected theorem map_iff [BorelSpace α] [MeasurableSpace β] [TopologicalSpace
+    {f : α → β} (hf : IsOpenEmbedding f) : (μ.comap f).Regular := by
+  haveI := OuterRegular.comap_of_continuous_of_measurableEmbedding
+    μ hf.continuous hf.measurableEmbedding
+  haveI :=
+    IsFiniteMeasureOnCompacts.comap' (μ := μ) hf.continuous hf.measurableEmbedding
```
```suggestion
  haveI := IsFiniteMeasureOnCompacts.comap' (μ := μ) hf.continuous hf.measurableEmbedding
```

EtienneC30 reviewed (commented):
Here are some more comments.

DavidLedvinka commented:
Ok I think I addressed all the comments. I also changed all of the theorems to (consistently) have the measure be an implicit parameter. But now I am having doubts that the measure should be explicit parameter. I can see arguments for both. What do you think? (Note map sometimes uses explicit and sometimes implicit so I think I originally used that as a guideline but its not clear those were intentional choices)

EtienneC30 commented on Mathlib/MeasureTheory/Constructions/BorelSpace/Basic.lean:
```diff
@@ -697,3 +697,28 @@ instance EReal.borelSpace : BorelSpace EReal :=
+
+variable [TopologicalSpace α] [TopologicalSpace β]
+
+namespace IsFiniteMeasureOnCompacts
```
I answered to your previous comment about `IsFiniteMeasureOnCompacts.comap'` and I agree it should indeed go to `Typeclasses.Finite`, I indicated a location in my answer. So I think these two small omits are ok. Another thing you could do (I find in a bit nicer but it's up to you) would be to name the typeclass parameters `[mα : MeasurableSpace]`, and then instead of
```lean
omit [MeasurableSpace α] in
```
you can write
```lean
variable {mα} in
```
 Of course the ideal situation would be to have `MeasurableSpace` implicit in the whole file, but it's not really important and definitely not in the scope of the PR.

EtienneC30 commented:
I answered to two of your remarks, they're the only unresolved conversations. Regarding the impliciteness of the measure, in my experience there is no rule for it and it seems a bit random. I think I have much more often be annoyed by explicit measures which can be inferred by unification than implicit ones having to be specified by hand.

On the other hand most of your statements are about typeclasses which therefore have to be written ad hoc with `have` because tactics won't generate side goals for them, so it's a matter of whether you prefer to write `have : Foo (μ.comap) f := Foo.comap hf` or `have := Foo.comap (μ := μ) hf`, or if you prefer to write `have := Foo.comap μ hf`. I tend to prefer the latter, so my conclusion would be to make it explicit if the conclusion is a class, and implicit otherwise.

This is my opinion but feel to do it as you want.

DavidLedvinka commented:
Ok I switched everything from implicit to explicit except for `InnerRegularWRT`

EtienneC30 reviewed (commented):
Two last tiny details. Thank you!
maintainer delegate

RemyDegenne commented:
Thanks!
bors r+
